radix-engine/src/model/resource_manager.rs

```diff
@@ -1,209 +1,102 @@
-        &["update_non_fungible_mutable_data"],
-    ),
-];
-
 impl ResourceManager {
     pub fn new(
         resource_type: ResourceType,
         metadata: HashMap<String, String>,
-        flags: u64,
-        mutable_flags: u64,
-        authorities: HashMap<ResourceAddress, u64>,
+        auth: ComponentAuthorization,
     ) -> Result<Self, ResourceManagerError> {
-        if !resource_flags_are_valid(flags) {
-            return Err(ResourceManagerError::InvalidResourceFlags(flags));
+        let mut authorization: HashMap<String, MethodAuthorization> = HashMap::new();
+        if let Some(mint_auth) = auth.get("mint") {
+            // TODO: Check for other invalid mint permissions?
+            if let MethodAuth::AllowAll = mint_auth {
+                return Err(ResourceManagerError::InvalidMintPermission);
+            }
+
+            authorization.insert(
+                "mint".to_string(),
+                convert(&Type::Unit, &Value::Unit, mint_auth),
+            );
         }
 
-        if !resource_flags_are_valid(mutable_flags) {
-            return Err(ResourceManagerError::InvalidResourceFlags(mutable_flags));
+        if let Some(burn_auth) = auth.get("burn") {
+            authorization.insert(
+                "burn".to_string(),
+                convert(&Type::Unit, &Value::Unit, burn_auth),
+            );
         }
 
-        let mut method_states: HashMap<String, MethodState> = HashMap::new();
-        method_states.insert(
-            "mint".to_string(),
-            MethodState::new(IsSet(MINTABLE), AlwaysTrue),
-        );
-        method_states.insert(
-            "burn".to_string(),
-            MethodState::new(IsSet(BURNABLE), IsNotSet(FREELY_BURNABLE)),
-        );
-        method_states.insert(
-            "take_from_vault".to_string(),
-            MethodState::new(AlwaysTrue, IsSet(RESTRICTED_TRANSFER)),
-        );
-        method_states.insert(
-            "enable_flags".to_string(),
-            MethodState::new(AlwaysTrue, AlwaysTrue),
-        );
-        method_states.insert(
-            "disable_flags".to_string(),
-            MethodState::new(AlwaysTrue, AlwaysTrue),
-        );
-        method_states.insert(
-            "lock_flags".to_string(),
-            MethodState::new(AlwaysTrue, AlwaysTrue),
-        );
-        method_states.insert(
-            "update_metadata".to_string(),
-            MethodState::new(IsSet(SHARED_METADATA_MUTABLE), AlwaysTrue),
-        );
-        method_states.insert(
-            "update_non_fungible_mutable_data".to_string(),
-            MethodState::new(IsSet(INDIVIDUAL_METADATA_MUTABLE), AlwaysTrue),
-        );
+        if let Some(take_auth) = auth.get("take_from_vault") {
+            authorization.insert(
+                "take_from_vault".to_string(),
+                convert(&Type::Unit, &Value::Unit, take_auth),
+            );
+        } else {
+            return Err(ResourceManagerError::TakeFromVaultNotDefined);
```

Comment:
Do we really need this check? This looks strange as we're taking a generic approach but paternalistic about this specific method. Any thoughts?

Reply:
Agreed it's weird, maybe we move the paternalism to the builder side while keeping system open?

Reply:
Yeah, would prefer that.